crates/shared/src/price_estimation/native/coingecko.rs

```diff
@@ -71,101 +72,167 @@ impl CoinGecko {
+                    .map(|token| format!("{:#x}", token))
+                    .collect::<Vec<_>>()
+                    .join(","),
+            )
             .append_pair("vs_currencies", "eth")
             .append_pair("precision", "full");
+
         let mut builder = self.client.get(url.clone());
         if let Some(ref api_key) = self.api_key {
             builder = builder.header(Self::AUTHORIZATION, api_key)
         }
         observe::coingecko_request(&url);
-        let response = builder.send().await.map_err(|e| {
-            PriceEstimationError::EstimatorInternal(anyhow!(
-                "failed to sent CoinGecko price request: {e:?}"
-            ))
-        })?;
+        let response = builder
+            .send()
+            .await
+            .context("failed to sent CoinGecko price request")?;
         if !response.status().is_success() {
-            let status = response.status();
-            return match status {
+            return match response.status() {
                 StatusCode::TOO_MANY_REQUESTS => Err(PriceEstimationError::RateLimited),
-                status => Err(PriceEstimationError::EstimatorInternal(anyhow!(
-                    "failed to retrieve prices from CoinGecko: error with status code {status}."
-                ))),
+                status => Err(PriceEstimationError::EstimatorInternal(anyhow!(format!(
+                    "CoinGecko returned non-success status code: {status}"
+                )))),
             };
         }
         let response = response.text().await;
         observe::coingecko_response(&url, response.as_deref());
-        let response = response.map_err(|e| {
-            PriceEstimationError::EstimatorInternal(anyhow!(
-                "failed to fetch native CoinGecko prices: {e:?}"
-            ))
-        })?;
-        let prices = serde_json::from_str::<Response>(&response)
-            .map_err(|e| {
-                PriceEstimationError::EstimatorInternal(anyhow!(
-                    "failed to parse native CoinGecko prices from {response:?}: {e:?}"
-                ))
-            })?
-            .0;
-
-        let price = prices
-            .get(&token)
-            .ok_or(PriceEstimationError::NoLiquidity)?;
-        Ok(price.eth)
+        let response = response.context("failed to fetch response body")?;
+
+        let parsed_response = serde_json::from_str::<Response>(&response)
+            .with_context(|| format!("failed to parse response: {response:?}"))?
+            .0
+            .into_iter()
+            .map(|(token, price)| (token, price.eth))
+            .collect();
+
+        Ok(parsed_response)
+    }
+
+    /// Fetches the prices of the given tokens denominated in the given token.
+    async fn bulk_fetch_denominated_in_token(
+        &self,
+        mut tokens: Vec<&Token>,
+        denominator: Token,
+    ) -> Result<HashMap<H160, NativePriceEstimateResult>, PriceEstimationError> {
+        tokens.push(&denominator);
```

Comment:
How do we make sure we don't make the list of tokens too large? Careful configuration of the batcher (ie. for xDAI max batch size needs to be 19 instead of 20?).

Alternatively, the `NativePriceBatchFetcher` trait could expose a method `batchSize` which would allow to colocate the logic of how big a batch can be with the implementation of the request (and here we could depending on the weth token symbol return 19 or 20 batch size.

Reply:
> Alternatively, the NativePriceBatchFetcher trait could expose a method batchSize

I played around with exactly that and it didn't seem too great given that we already have this as a config value.

Reply:
If `tokens` is 20, then this query would fail, because appending extra token will make CoinGecko query invalid (I don't know how it behaves in that case, if it ignores the extra token or directly fails).